sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubConsumerClient.cs

```diff
@@ -417,6 +418,171 @@ await foreach (var partitionEvent in eventChannel.Reader.EnumerateChannel(option
+                                                                        CancellationToken cancellationToken = default) => ReadEventsAsync(true, readOptions, cancellationToken);
+
+        /// <summary>
+        ///   Reads events from all partitions of the event hub as an asynchronous enumerable, allowing events to be iterated as they
+        ///   become available on the partition, waiting as necessary should there be no events available.
+        ///
+        ///   This enumerator may block for an indeterminate amount of time for an <c>await</c> if events are not available on the partition, requiring
+        ///   cancellation via the <paramref name="cancellationToken"/> to be requested in order to return control.  It is recommended to set the
+        ///   <see cref="ReadOptions.MaximumWaitTime" /> for scenarios where a more deterministic maximum waiting period is desired.
+        /// </summary>
+        ///
+        /// <param name="startReadingAtEarliestEvent"><c>true</c> to begin reading at the first events available in each partition; otherwise, reading will begin at the end of each partition seeing only new events as they are published.</param>
+        /// <param name="readOptions">The set of options to use for configuring read behavior; if not specified the defaults will be used.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>An <see cref="IAsyncEnumerable{T}"/> to be used for iterating over events in the partition.</returns>
+        ///
+        /// <remarks>
+        ///   This method is not recommended for production use; the <see cref="EventProcessorClient" /> should be used for reading events from all partitions in a
+        ///   production scenario, as it offers a much more robust experience with higher throughput.
+        ///
+        ///   It is important to note that this method does not guarantee fairness amongst the partitions during iteration; each of the partitions competes to publish
+        ///   events to be read by the enumerator.  Depending on service communication, there may be a clustering of events per partition and/or there may be a noticeable
+        ///   bias for a given partition or subset of partitions.
+        ///
+        ///   Each reader of events is presented with an independent iterator; if there are multiple readers, each receive their own copy of an event to
+        ///   process, rather than competing for them.
+        /// </remarks>
+        ///
+        /// <seealso cref="EventProcessorClient" />
+        /// <seealso cref="ReadEventsAsync(CancellationToken)"/>
+        /// <seealso cref="ReadEventsAsync(ReadOptions, CancellationToken)"/>
+        ///
+        public virtual async IAsyncEnumerable<PartitionEvent> ReadEventsAsync(bool startReadingAtEarliestEvent,
```

Comment:
Hi,
Quick question: Why not take the startingPosition as parameter directly?
For example, that way, the consumer can choose to resume reading from wherever it stopped.

Reply:
Hi @KPixel,

Thanks for your question.  

The `ReadEvents` family of methods are intended to be used for exploring Event Hubs and getting started with a fast developer iteration loop.   Because of that, the method is intended to allow use without having to explicitly understand the partitions of the associated Event Hub nor expose the complexity of having to understand and choose an explicit starting position for each.

We want to discourage using `ReadEvents` for production scenarios and guide developers to the `EventProcessorClient`, which is intended to be a more performant and robust client for reading events across partitions.  For `EventProcessorClient`, you are able to provide an explicit default starting position for each partition, as well as having partitions treated with equal distribution rather than competing to emit events as the `ReadEvents` methods do.

Reply:
I agree with discouraging using `ReadEvents` for production, and I would then suggest even moving it outside, maybe as an extension in a clearly labelled class (a less extreme version of what you did with the in-memory checkpoint store).
But I would understand if you don't want to go that far.

Reply:
Unfortunately, while I like the idea, I feel that it would work against the goal of having it discoverable and easily accessible for those unfamiliar with Event Hubs.

Your description of "layers" is pretty much spot-on for the direction that we're trying to go.   We're trying to optimize the ease of use and discoverability for our "Hello World" scenarios on the main client classes themselves, while providing as clear and easy a "step-up" path as we're able for real-world production scenarios, typically involving moving to a more specialized set of client types.  

As you've mentioned, there's bound to be some tension between them, as accentuating the easier experiences on the client and trying to keep complexity reduced makes it harder to discover the more advanced types and has potential to allow developers to feel complacent with the "Hello World" surface.  Unfortunately, there's no clear and effective winning approach to balance that to serve all of our scenarios; there's just the expected set of trade-offs.

We're hoping that we'll be able to help guide people towards the pit of success through choosing to intentionally limit some scenarios, have a comprehensive set of samples that show a "step-up" progression, and document the heck out of things.  We're obviously not quite in our end state yet, but those are our goals.

If you're interested in some context around the guidelines that we're using for the SDKs, they're documented in the [Azure SDK General Guidelines](https://azure.github.io/azure-sdk/general_introduction.html) and [Azure SDK Guidelines for .NET](https://azure.github.io/azure-sdk/dotnet_introduction.html).

We very much appreciate feedback and would welcome your perspective and discussion - especially around those scenarios where you see gaps that we may have overlooked or have ideas how to tune that balance.  You're welcome to open issues for discussion, either in this repository for .NET-specific topics or in the [Azure SDK Repository](https://github.com/Azure/azure-sdk) for those that are more general or related to the guidelines.